### ISSUE TITLE
Weight Standardization

### ISSUE THREAD

**Issue by KeisukeShimokawa** [KeisukeShimokawa/papers-challenge#117]
### 論文へのリンク

[[arXiv:1903.10520] Weight Standardization](arXiv:1903.10520)

### 著者・所属機関

Siyuan Qiao, Huiyu Wang, Chenxi Liu, Wei Shen, Alan Yuille

- Johns Hopkins University

### 投稿日時（YYYY-MM-DD）

2019-05-25

**Comment by KeisukeShimokawa** [comment]
[https://github.com/joe-siyuan-qiao/pytorch-classification/tree/e6355f829e85ac05a71b8889f4fff77b9ab95d0b](https://github.com/joe-siyuan-qiao/pytorch-classification/tree/e6355f829e85ac05a71b8889f4fff77b9ab95d0b)

**Comment by KeisukeShimokawa** [comment]
公式実装より

```python
class Conv2d(nn.Conv2d):

    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                 padding=0, dilation=1, groups=1, bias=True):
        super(Conv2d, self).__init__(in_channels, out_channels, kernel_size, stride,
                 padding, dilation, groups, bias)

    def forward(self, x):
        # return super(Conv2d, self).forward(x)
		
		# [Cout, Cin, K, K]
        weight = self.weight
        # Normalize [Cin, K, K] => [Cout, 1, 1, 1]
		weight_mean = weight.mean(dim=1, keepdim=True).mean(dim=2,
                                  keepdim=True).mean(dim=3, keepdim=True)
		# [Cout, Cin, K, K]
        weight = weight - weight_mean
		# std(dim=1) => [Cout, CinxKxK] => [Cout, 1, 1, 1]
        std = weight.view(weight.size(0), -1).std(dim=1).view(-1, 1, 1, 1) + 1e-5
		# std.shape [Cout, 1, 1, 1] => [Cout, Cin, K, K]
        weight = weight / std.expand_as(weight)
		# use this weight
        return F.conv2d(x, weight, self.bias, self.stride,
                        self.padding, self.dilation, self.groups)


def BatchNorm2d(num_features):
	# channels group 32 is default
    return nn.GroupNorm(num_channels=num_features, num_groups=32)
```

**Comment by KeisukeShimokawa** [comment]
各種正規化手法で計算するモーメントの対象

![image](https://user-images.githubusercontent.com/33191285/82080606-d9ad1200-971f-11ea-9357-050d83d4b3e1.png)

Weight Normalizationで正規化を行う対象は以下になる。

![image](https://user-images.githubusercontent.com/33191285/82080636-e7fb2e00-971f-11ea-9d9c-123a01ae546d.png)

**Comment by KeisukeShimokawa** [comment]
- [37] S. Santurkar, D. Tsipras, A. Ilyas, and A. Madry. How does
batch normalization help optimization? In Advances in Neural Information Processing Systems (NeurIPS), pages 2488–
2498, 2018. 1, 2, 8
